### PR TITLE
Do not offer UseObjectInitializer if it would change implicit-member-access semantics.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/15012

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
As part of this i changed the behavior of several syntax facts methods to be far more explicit than implicit.  SyntaxFacts are generally used to do very simply syntactic operations in a VB/C# agnostic fashion.  like "Give me the name of the member access expression i have".   

While higher level functionality can be provided (for example, functions that smooth out hte differences between ```.``` vs ```?.```) those should be in helpers that clearly state that they're doing that smoothing out).  Otherwise, code that does not want special ```?.``` handling now needs to realize that it must opt-out of this behavior.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

File: src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb [hunk 1]
```diff
@@ -184,6 +184,50 @@ Class C
...
 
+        <WorkItem(15012, "https://github.com/dotnet/roslyn/issues/15012")>
+        <Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseObjectInitializer)>
+        Public Async Function TestMissingIfImplicitMemberAccessWouldChange() As Task
+            Await TestMissingAsync(
+"
+Class C
+    Sub M()
+        With New String()
+            Dim x As ProcessStartInfo = [||]New ProcessStartInfo()
```
Comment (dpoeschl) [inline src/EditorFeatures/VisualBasicTest/UseObjectInitializer/UseObjectInitializerTests.vb]: What if you're in a function inside of a With (resetting the With boundary)?

File: src/Features/Core/Portable/UseObjectInitializer/AbstractUseObjectInitializerDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -293,6 +306,38 @@ internal struct Analyzer<
...
 
+        private bool ImplicitMemberAccessWouldBeAffected(SyntaxNode node)
+        {
+            if (node != null)
+            {
+                foreach (var child in node.ChildNodesAndTokens())
+                {
+                    if (child.IsNode)
+                    {
+                        if (ImplicitMemberAccessWouldBeAffected(child.AsNode()))
```
Comment (dpoeschl) [inline src/Features/Core/Portable/UseObjectInitializer/AbstractUseObjectInitializerDiagnosticAnalyzer.cs]: Combine with outer if?

File: src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs [hunk 3]
```diff
@@ -561,12 +553,11 @@ public void GetNameAndArityOfSimpleName(SyntaxNode node, out string name, out in
...
-        public SyntaxNode GetExpressionOfMemberAccessExpression(SyntaxNode node)
-        {
-            return node.IsKind(SyntaxKind.MemberBindingExpression)
-                ? GetExpressionOfConditionalAccessExpression(node.GetParentConditionalAccessExpression())
-                : (node as MemberAccessExpressionSyntax)?.Expression;
-        }
+        public SyntaxNode GetTargetOfMemberBinding(SyntaxNode node)
+            => (node as MemberBindingExpressionSyntax).GetParentConditionalAccessExpression()?.Expression;
+
+        public SyntaxNode GetExpressionOfMemberAccessExpression(SyntaxNode node, bool allowImplicitTarget)
```
Comment (dpoeschl) [inline src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs]: Are all the callers of this safe?

**Pull request merged**